---
title: "F2 simulations results"
date: "`r Sys.Date()`"
author: "[Statistical Genetics Lab](http://statgen.esalq.usp.br) <br/> Department of Genetics <br/> Luiz de Queiroz College of Agriculture <br/> University of São Paulo"
output:
    rmdformats::readthedown:
      css: readthedownstatgen.css
      code_folding: hide
---

```{r knitr_init, echo=FALSE, cache=FALSE}
library(knitr)
library(rmdformats)

## Global options
options(max.print="75")

opts_knit$set(width=75)
```

This report include analysis with all output files generated by the simulations workflow for F2 populations.

## Output files

Short description about the output files.

From pedsim_files task:

* mapfile.map

marker chromosome position(cM)

* tot_mks.txt

chr pos ref alt 

From aval_vcf task:

* method_alt_depth.txt

method = c("gatk", "freebayes", "stacks")

columns = individuals
lines = markers
cells = counts for alternative alleles

* method_ref_depth.txt
																																																																						
method = c("gatk", "freebayes", "stacks")

columns = individuals
lines = markers
cells = counts for reference alleles

* method_GQ.txt

Ta com pau.. arrumar!

* method.txt
nmk.filt nmk.id ok falso.positivo ref.ok alt.ok

From all_maps task:

* method_error_info_genotypeMethod.txt

method = c("gatk", "freebayes", "stacks")
genotypeMethod = c("GQ", "polyrad", "supermassa", "updog")

MK POS gabGT methGT 1 2 3 4 5

* method_filters_genotypeMethod.txt

method = c("gatk", "freebayes", "stacks")
genotypeMethod = c("dfAndGQ", "polyrad", "supermassa", "updog")

n_markers distorted_markers redundant_markers

* method_map_genotypeMethod.txt

method = c("gatk", "freebayes", "stacks")
genotypeMethod = c("df", "GQ", "polyrad", "supermassa", "updog")

MK POS cM


# Path

**The only part of this report that you need to change**

```{r, warning=FALSE, message=FALSE, results='hide'}
path <- "~/github/errors_workflow/cromwell-executions/F2/0053b719-69f8-466c-b3f0-861f5a7c35d1/"
cMByMb <- 4.63
```


# Packages

```{r, warning=FALSE, message=FALSE, results='hide'}
library(ggplot2)
library(reshape2)
library(vcfR)
```

# Code for only one family - pilot

```{r, warning=FALSE, message=FALSE, results='hide'}
setwd(path)
method <- c("gatk", "freebayes")
meth.geno <- c("GQ", "polyrad", "updog", "supermassa")

mapfile <- read.table("call-pedsim_files/execution/mapfile.map", header=T)

tot_mks <- read.table("call-pedsim_files/execution/tot_mks.txt")
id <- c(0,1)
p  <- list()


df.diff.dis <- as.list(rep(NA, length(method)))
for(i in 1:length(df.diff.dis)){
  df.diff.dis[[i]] <- as.list(rep(NA, length(meth.geno)))
}

df.per.geno <- df.coverage <- df.diff.tot <- df.diff.median <- df.diff.mean <- names.id <- vector()

for(i in 1:length(method)){
  ## vcfs
  vcf <- read.vcfR(paste0("call-vcftools_filter/execution/",method[i], ".recode.vcf"))
  gt <- vcf@gt[,-1]
  gt <- melt(gt)
  gt$value <- sapply(strsplit(as.character(gt$value), ":"), "[", 1)
  
  ## Depths
  alt_depth <- read.table(paste0("call-aval_vcf/execution/", method[i], "_alt_depth.txt"), header = T, stringsAsFactors = F)
  alt_depth <- as.data.frame(apply(alt_depth, 2, as.integer))
  alt <- melt(alt_depth)
  
  ref_depth <- read.table(paste0("call-aval_vcf/execution/", method[i], "_ref_depth.txt"), header = T, stringsAsFactors = F)
  ref_depth <- as.data.frame(apply(ref_depth, 2, as.integer))
  ref <- melt(ref_depth)
  df.depths <- data.frame(ind = gt$Var2, gt = gt$value, alt = alt$value, ref=ref$value)
  
  p[[i]] <- ggplot(df.depths, aes(x=ref, y=alt, color=as.factor(gt))) + geom_point(size=0.5) +
    labs(title= paste0(method[i]," depths example"),x="ref", y = "alt", color="Genotypes") 
  
  for(j in 1:length(meth.geno)){
    # Map distance
    maps <- read.table(paste0("call-all_maps/shard-", id[i],"/execution/", method[i],"_map_", meth.geno[j],".txt"), header =T)
    
    poscM <- (as.numeric(as.character(maps[,2]))/1000000)*cMByMb
    poscM.norm <- poscM-poscM[1]
    maps <- cbind(maps, poscM, poscM.norm)
    
    # Difference between distances real and estimated
    diff.dis <- sqrt((maps$poscM.norm - maps$rf)^2)
    df.diff.dis[[i]][[j]] <- diff.dis
    names(df.diff.dis[[i]][[j]]) <- paste0(method[i],"_",meth.geno[j])
    
    # Meth names
    names.id <-  c(names.id,paste0(method[i],"_",meth.geno[j]))
    
    # Mean of differences
    diff.dis.mean <- mean(sqrt((maps$poscM.norm - maps$rf)^2))
    df.diff.mean <- c(df.diff.mean,diff.dis.mean)
    
    
    # Median of differences
    diff.dis.median <- median(sqrt((maps$poscM.norm - maps$rf)^2))
    df.diff.median <- c(df.diff.median,diff.dis.median)
    
    # Difference of total size
    diff.tot.dis <- sqrt((maps$poscM.norm[length(maps$poscM.norm)] - maps$rf[length(maps$rf)])^2)
    df.diff.tot <- c(df.diff.tot,diff.tot.dis)
    
    mapfile.pos <- cbind(mapfile, tot_mks[,2])
    
    # Percentage of the chromosome covered 
    coverage <- maps$pos[length(maps$pos)]*100/mapfile.pos[,4][length(mapfile.pos[,4])]
    df.coverage <- c(df.coverage,coverage)
    
    # Percentage of rigth genotyping
    per.geno <- read.table(paste0("call-all_maps/shard-",id[i],"/execution/",method[i],"_error_info_",meth.geno[j],".txt"), header =T)
    per.geno <- (sum(per.geno$gabGT == per.geno$methGT)/length(per.geno$methGT)*100)
    df.per.geno <- c(df.per.geno,per.geno)
    
  }              
}


```

## Depth

Aqui mostra a distribuição das contagens de cada alelo, teoricamente deveriam ter homozigotos para os dois pais e heterozigotos no meio, uma diagonal (mas parece que só esta simulando homozigotos, isso altera tudo em seguida).


```{r}
# Depths
p
```


##  Difference between distances real and estimated

As distâncias estimadas do mapa deveriam ser as mesmas das simuladas, quanto mais perto do 0 o valor, melhor. O primeiro número sempre vai ser 0 porque eu normalizei.

```{r}
df.diff.dis
```

## Difference of total size

Como serão várias famílias, vai ficar difícil comparar em relação à todas as marcas, então estou considerando comparar só o tamanho total, média e mediana de cada família

```{r}
rbind(names.id,df.diff.tot)
```

## Median of differences

```{r}
rbind(names.id,df.diff.median)
```

## Mean of differences

```{r}
rbind(names.id,df.diff.mean)
```

## Percentage of rigth genotyping

Comparação entre os genótipos simulados e estimados com cada método

```{r}
rbind(names.id,df.per.geno)
```

## Percentage of the chromosome covered 

Aqui esta outro problema, ele esta simulando somente para parte do genoma

```{r}
rbind(names.id,df.coverage)
```

```
Só para lembrar:
Acho que rola fazer uma discussão sobre numero de indivíduos e resolução do mapa, com poucos individuos o mapa fica inflado tb

Rola fazer outra discussão sobre a cobertura que o RADseq cobre , a capacidade dos softwares de identificar no genoma inteiro - Se o tamanho total do mapa nao corresponder ao cromossomo inteiro pode ser que ele esteja inflado e as pessoas pensem que ele esta no tamanho certo
``` 
