---
title: "F2 simulations results"
date: "`r Sys.Date()`"
author: "[Statistical Genetics Lab](http://statgen.esalq.usp.br) <br/> Department of Genetics <br/> Luiz de Queiroz College of Agriculture <br/> University of São Paulo"
output:
    rmdformats::readthedown:
      css: readthedownstatgen.css
---

```{r knitr_init, echo=FALSE, cache=FALSE}
library(knitr)
library(rmdformats)

## Global options
options(max.print="75")

opts_knit$set(width=75)
```

This report include analysis with all output files generated by the simulations workflow for F2 populations.

**Output files**

From pedsim_files task:

* mapfile.map

marker chromosome position(cM)

* tot_mks.txt

chr pos ref alt 

From aval_vcf task:

* method_alt_depth.txt

method = c("gatk", "freebayes", "stacks")

columns = individuals
lines = markers
cells = counts for alternative alleles

* method_ref_depth.txt

method = c("gatk", "freebayes", "stacks")

columns = individuals
lines = markers
cells = counts for reference alleles

* method_GQ.txt

Ta com pau.. arrumar!


* method.txt
nmk.filt nmk.id ok falso.positivo ref.ok alt.ok

From all_maps task:

* method_error_info_genotypeMethod.txt

method = c("gatk", "freebayes", "stacks")
genotypeMethod = c("GQ", "polyrad", "supermassa", "updog")

MK POS gabGT methGT 1 2 3 4 5

* method_filters_genotypeMethod.txt

method = c("gatk", "freebayes", "stacks")
genotypeMethod = c("dfAndGQ", "polyrad", "supermassa", "updog")

n_markers distorted_markers redundant_markers

* method_map_genotypeMethod.txt

method = c("gatk", "freebayes", "stacks")
genotypeMethod = c("df", "GQ", "polyrad", "supermassa", "updog")

MK POS cM


# Path

**The only part of this report that you need to change**

The path of the workflow results.

```{r}
path <- ""
```


# Packages

```{r}
library(ggplot2)
library(reshape2)
```

Acho que rola fazer uma discussão sobre numero de indivíduos e resolução do mapa, com poucos individuos o mapa fica inflado tb

Rola fazer outra discussão sobre a cobertura que o RADseq cobre , a capacidade dos softwares de identificar no genoma inteiro - Se o tamanho total do mapa nao corresponder ao cromossomo inteiro pode ser que ele esteja inflado e as pessoas pensem que ele esta no tamanho certo

```{r}

# Graphics

maps.names <- c("map.df",
                "map.up",
                "map.poly",
                "map.super")

# Total size
genotype <- rep(maps.names, each=n.sim)

tot.size <- list()
for(i in 1:length(maps.names))
  tot.size[[i]] <- sapply(get(maps.names[i])[[3]], function(x) x[length(x)])
tot.size <- unlist(tot.size)

tot.size.df <- data.frame(genotype, tot.size)

p <- ggplot(tot.size.df, aes(x=genotype, y=tot.size)) + geom_boxplot() +
  labs(title= paste0("Total size ", depth[j]) ,x="genotype and probs", y = "cM")   + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1), 
        panel.background = element_rect(fill = "lightblue",
                                        colour = "lightblue",
                                        size = 0.5, linetype = "solid")) +
  geom_hline(yintercept = 12, color = "darkgreen") 



ggsave(p,filename = paste0("Tot.size.",depth[j],".jpg"))
p

# Distances between markers

df.dist.part <- data.frame()
for(i in 1:length(maps.names)){
  dist.part <- unlist(get(maps.names[i])[[2]])
  df.dist.part <- rbind(df.dist.part, data.frame(geno=rep(maps.names[i], length(dist.part)), dist = dist.part))
}


p <- ggplot(df.dist.part, aes(x=geno, y=dist)) + geom_boxplot() +
  labs(title= paste0("Distances ", depth[j]) ,x="genotype and probs", y = "cM")   + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1), 
        panel.background = element_rect(fill = "lightblue",
                                        colour = "lightblue",
                                        size = 0.5, linetype = "solid")) +
  geom_hline(yintercept = 1, color = "darkgreen") 

ggsave(p,filename = paste0("Dists.",depth[j],".jpg"))
p
```





