---
title: "Empiricalreads workflow"
date: "`r Sys.Date()`"
author: "[Statistical Genetics Lab](http://statgen.esalq.usp.br) <br/> Department of Genetics <br/> Luiz de Queiroz College of Agriculture <br/> University of SÃ£o Paulo"
output:
    rmdformats::readthedown:
      css: readthedownstatgen.css
---


* EmpiricalReads.wdl

This one has as input the demultiplexed and filtered fastq files and the reference genome. The fastq files are aligned to the reference genome, the SNP calling is performed with gatk and freebayes software, the genotype calling is made with updog, polyrad and supermassa software and linkage maps are built with gusmap and onemap.

Sub-workflows are available for each step of the main workflows:

input: reference genome
output: fastq files

![](images/EmpiricalReads.svg)

* Create_aligments_from_families_files

input: fastq files
output: bam files

![](images/create_alignment_from_families_files.svg)

* freebayes_genotyping.wdl

input: bam files
output: vcf file

![](images/freebayes_genotyping.svg)

* gatk_genotyping.wdl

input: bam files
output: vcf file

![](images/gatk_genotyping.svg)

* default_maps.wdl

input: vcf files
output: RData and reports

![](images/default_maps_emp.svg)

* SNPCaller_maps.wdl

input: vcf files
output: RData and reports

![](images/snpcaller_maps_emp.svg)

* polyrad_maps.wdl

input: vcf files
output: RData and reports

![](images/polyrad_maps_emp.svg)

* supermassa_maps.wdl

input: vcf files
output: RData and reports

![](images/supermassa_maps_emp.svg)

* updog_maps.wdl

input: vcf files
output: RData and reports

![](images/updog_maps_emp.svg)

* gusmap_maps.wdl

input: vcf files
output: RData and reports

![](images/gusmap_maps_emp.svg)

You can also build your own workflow using tasks available in utils.wdl and utilsR.wdl
