---
title: "High-Density Maps"
date: "`r Sys.Date()`"
author: "[Statistical Genetics Lab](http://statgen.esalq.usp.br) <br/> Department of Genetics <br/> Luiz de Queiroz College of Agriculture <br/> University of SÃ£o Paulo"
output:
    rmdformats::readthedown:
      css: readthedownstatgen.css
---

This tutorial will help you to build linkage maps using sequencing technologies. If you have in hands:

* An outcrossing diploid specie
* Bi-parental cross
* Fastq or VCF files
* Reference genome of the specie or some related

And wants to build linkage map for the population, this tutorial is for you!

Since version 3.0, OneMap can use genotype probabilities, instead of just genotypes, to estimate the genetic distances. This had shown to be a very good idea to control genotyping errors, which use to inflate the map sizes. For more information about this, please, see [this paper]().

OneMap 3.0 can also use informations from Multiallelic Nucleotide Polymorphisms (MNPs) identified by Freebayes or GATK SNP callers. Further information in [this paper]().

Before we start, one question that we also had after implementing the genotypes probabilities was "Which one to use?", once there are several genotype calling software available. To help answer this question, we made a workflow that uses probabilities from different softwares: Freebayes, GATK, Updog, Supermassa and PolyRAD. At the end, you will have one linkage group built with each combination of SNP, genotype method and probability used. After selecting your best probabilities avaliating the generated graphics in a shiny app, you can export the onemap object choosed and use the same methods to the rest of the groups. See the examples in our [shiny app](https://github.com/Cristianetaniguti/onemap_workflows_shiny).

We know that there are more software available that perfom the SNP and genotype calling and our workflow is easily expansible, we were very curious to test others, but we had to stop in some point. 

It is also important to highlight that we use default paramenters of all software implemented in the workflow. But you can change any of them just modifying the scripts (you will find where easily, don't worry, [see this tutorial](https://github.com/Cristianetaniguti/onemap_workflows))

If you did your SNP calling in other software, pay attention if your VCF has the GQ field, we can use this value as genotype error. If your VCF have also the allele depth information you can still use the workflow and test the genotype callers Updog, Supermassa and PolyRAD. 

We worked hard to make this workflow very user-friendly. Then you can run all programs and install just two of them: [java](https://www.java.com/en/) and  **[Docker](https://www.docker.com/)**.

The workflow is very useful but also very heavy in terms of computational demand. We expect that you have access to a cluster to be able to run it. We include special configurations for HPC with [slurm](https://slurm.schedmd.com/documentation.html) system and with [singularity](https://singularity.lbl.gov/docs-docker) program available. Singularity allows you to use Docker containers without the super user permission. If the HPC you use do not have it installed, it would be a good ideia to ask to the administrator to install.

Before run, you can have an idea of how much memory and time will be needed by the workflow for one of your groups. Access [here]().

Okay, now we are ready! Go to the session Run with Empirical Data in [github repository with the workflow](https://github.com/Cristianetaniguti/onemap_workflows) to know how you can run it for your data.

